src/Cli/dotnet/commands/dotnet-tool/install/ToolInstallLocalCommand.cs

```diff
@@ -46,39 +54,58 @@ internal class ToolInstallLocalCommand : CommandBase
             _toolManifestEditor = toolManifestEditor ?? new ToolManifestEditor();
             _localToolsResolverCache = localToolsResolverCache ?? new LocalToolsResolverCache();
-            _toolLocalPackageInstaller = new ToolInstallLocalInstaller(parseResult, toolPackageDownloader);
+            _toolLocalPackageInstaller = new ToolInstallLocalInstaller(parseResult, _packageId, toolPackageDownloader);
+            _toolPackageDownloader = toolPackageDownloader;
             _allowRollForward = parseResult.GetValue(ToolInstallCommandParser.RollForwardOption);
-
-            _packageId = new PackageId(parseResult.GetValue(ToolUpdateCommandParser.PackageIdArgument));
             _allowPackageDowngrade = parseResult.GetValue(ToolInstallCommandParser.AllowPackageDowngradeOption);
-
         }
-        
 
         public override int Execute()
+        {
+            if (_all)
+            {
+                var toolListCommand = new ToolListLocalCommand(_parseResult, (IToolManifestInspector)_toolManifestFinder);
+                var toolIds = toolListCommand.GetPackages(null);
+                foreach (var toolId in toolIds)
+                {
+                    ExecuteInstallCommand(toolId.Item1.PackageId);
+                }
+                return 0;
+            }
+            else
+            {
+                return ExecuteInstallCommand((PackageId) _packageId);
+            }
+        }
+
+        private int ExecuteInstallCommand(PackageId packageId)
         {
             FilePath manifestFile = GetManifestFilePath();
 
             (FilePath? manifestFileOptional, string warningMessage) =
-                _toolManifestFinder.ExplicitManifestOrFindManifestContainPackageId(_explicitManifestFile, _packageId);
+                _toolManifestFinder.ExplicitManifestOrFindManifestContainPackageId(_explicitManifestFile, packageId);
 
             if (warningMessage != null)
             {
                 _reporter.WriteLine(warningMessage.Yellow());
             }
 
             manifestFile = manifestFileOptional ?? GetManifestFilePath();
-            var existingPackageWithPackageId = _toolManifestFinder.Find(manifestFile).Where(p => p.PackageId.Equals(_packageId));
+            var existingPackageWithPackageId = _toolManifestFinder.Find(manifestFile).Where(p => p.PackageId.Equals(packageId));
 
             if (!existingPackageWithPackageId.Any())
             {
-                return InstallNewTool(manifestFile);
+                return InstallNewTool(manifestFile, packageId);
             }
 
             var existingPackage = existingPackageWithPackageId.Single();
-            var toolDownloadedPackage = _toolLocalPackageInstaller.Install(manifestFile);
+            if (!packageId.Equals(_packageId))
+            {
+                _toolLocalPackageInstaller = new ToolInstallLocalInstaller(_parseResult, packageId, _toolPackageDownloader);
```

Comment:
Unfortunately, this may be a bug for ToolInstallLocalInstaller imo, without newing a new ToolInstallLocalInstaller, the old ToolInstallLocalInstaller will always install the old package even though you passed in a new package ID in `ToolInstallLocalInstaller.Install`

Reply:
I'm suggesting we could change `ToolInstallLocallInstaller.Install` so that it uses a parameter passed in instead of the `_packageId` field, similar to the way `ToolInstallGlobalOrToolPathCommand.ExecuteInstallCommand` now works.

Reply:
> I'm suggesting we could change `ToolInstallLocallInstaller.Install` so that it uses a parameter passed in instead of the `_packageId` field, similar to the way `ToolInstallGlobalOrToolPathCommand.ExecuteInstallCommand` now works.

Sorry for the misread. Done!